Objects/floatobject.c

```diff
@@ -1683,8 +1685,69 @@ typedef enum _py_float_format_type float_format_type;
 #define ieee_big_endian_format _py_float_format_ieee_big_endian
 #define ieee_little_endian_format _py_float_format_ieee_little_endian
 
-#define float_format (_PyRuntime.float_state.float_format)
-#define double_format (_PyRuntime.float_state.double_format)
+/* We attempt to determine if this machine is using IEEE
+    floating point formats by peering at the bits of some
+    carefully chosen values.  If it looks like we are on an
+    IEEE platform, the float packing/unpacking routines can
+    just copy bits, if not they resort to arithmetic & shifts
+    and masks.  The shifts & masks approach works on all finite
+    values, but what happens to infinities, NaNs and signed
+    zeroes on packing is an accident, and attempting to unpack
+    a NaN or an infinity will raise an exception.
+
+    Note that if we're on some whacked-out platform which uses
+    IEEE formats but isn't strictly little-endian or big-
+    endian, we will fall back to the portable shifts & masks
+    method.
+
+    These functions are reduced to a no-op via compiler
+    optimizations. */
```

Comment:
Nowadays we have Annex F.  Maybe we should first check if ``__STDC_IEC_559__`` defined, else (if it does make sense at all) - fallback to test functions.  We have now also defines like ``DOUBLE_IS_BIG_ENDIAN_IEEE754`` from the configure script.

@picnixz, do you think these tests actually belongs to runtime?

Reply:
My thinking is: whenever you can deduce those information in advance, do it. Now, I'm not sure whether we need to keep this kind of test at runtime in cross-compiled binaries where the host and the target machines have different configurations.


cc @corona10 for this question of cross-compilation.

Previously, we cached the result of `get_float_format` for subsequent calls to `float.__getformat__` so I don't really know why not caching it would be faster. Maybe it's because of the global runtime state?